Source/DafnyStandardLibraries/src/DafnyStdLibs/DynamicArray.dfy

```diff
@@ -0,0 +1,244 @@
+      Repr := {this, data};
+    }
+
+    /**
+    Retrieve the element at index 
+     */
+    function At(index: uint32) : (element: A)
+      reads this, this.data
+      requires index < size
+      requires Valid?()
+      ensures element == data[index] == items[index]
+    {
+      data[index]
+    }
+
+    /**
+    Put element in the array at index 
+     */
+    method Put(index: uint32, element: A)
+      requires index < size
+      requires Valid?()
+      modifies data, `items
+      ensures Valid?()
+      ensures items == old(items)[index := element]
+    {
+      data[index] := element;
+      items := items[index := element];
+    }
+
+    /**
+    Ensure that at least a reserved amount of elements can still be pushed onto the array in constant time
+
+    Returns false only if it was not at all possible to provide the ensurance.
+     */
+    method Ensure(reserved: uint32) returns (s: bool)
+      requires Valid?()
+      ensures Valid?()
+      modifies `capacity, `data, `Repr, data
+      ensures s == (reserved <= MAX_CAPACITY - size)
+      ensures s ==> reserved <= capacity - size
+      ensures var largeEnough := old(reserved <= capacity - size);
+              if !s || largeEnough
+              then unchanged(this) && unchanged(data)
+              else fresh(data)
+    {
+      if reserved > MAX_CAPACITY - size {
+        return false;
+      }
+      if reserved <= capacity - size {
+        return true;
+      }
+      var new_capacity := capacity;
+      while reserved > new_capacity - size
+        decreases MAX_CAPACITY - new_capacity
+        invariant new_capacity >= capacity
+      {
+        new_capacity := DefaultNewCapacity(new_capacity);
+      }
+      Realloc(new_capacity);
+      return true;
+    }
+
+    /**
+    Pop an element from the array in constant time
+     */
+    method PopFast()
+      requires Valid?()
+      requires size > 0
+      modifies `size, `items
+      ensures Valid?()
+      ensures size == old(size) - 1
+      ensures items == old(items[..|items| - 1])
+    {
+      size := size - 1;
+      items := items[..|items| - 1];
+    }
+
+    /**
+    Push an element onto the array in constant time
+     */
+    method PushFast(element: A)
+      requires Valid?()
+      requires size < capacity
+      modifies `size, `items, data
+      ensures Valid?()
+      ensures size == old(size) + 1
+      ensures items == old(items) + [element]
+    {
+      data[size] := element;
+      size := size + 1;
+      items := items + [element];
+    }
+
+    /**
+    Push an element onto the array in constant time if there is available capacity, 
+    and otherwise in time linear to the current size of the array.
+
+    Returns false only if it was not at all possible to push the element onto the array.
+     */
+    method Push(element: A) returns (s: bool)
+      requires Valid?()
+      modifies this, data
+      ensures Valid?()
+      ensures s == (old(size) < MAX_CAPACITY)
```

Comment:
If you change the `==` to a `<==>`, you can omit the parentheses (like you do on L205).